### DISCUSSION TITLE
Is it good approach if i create indexes  of every Category?

### DISCUSSION THREAD

**Discussion by Talhaz** [rom1504/clip-retrieval#49] [Q&A] [answered]
Let Suppose I have a **fashion clothing store**, and to get accurate results should I make **indexes** of each category like [Tops, Blouse, jeans] ?
Because I tested it, and I created a folder include every 200 images of different categories but it did not show good results to me 
So I thought I should make indexes of every category and after classification by _CNN model_ then I will pass the image to _clip backend_  
But in your case, the **OpenAI clip** has the same model that gives almost 99.9% accurate similar items, Can u Explain to me how it works guide me about it?

**Comment by rom1504** [comment, marked as answer]
can you explain what feature you want to implement?

**Reply by Talhaz** [reply]
I just want to know how your one model gives the most accurate similar images? But in my case when I made the index image folder which contain different kinds of images It did not give me correct similar images.
that's the problem so I thought i should run inference separately to search the image into its own category index file.

**Reply by Talhaz** [reply]
In a collab demo notebook we created an embeddings of a Cat dataset , but what happens if we add more categories like cats of different types , dogs and its breeds , clothes images  then should we run a inference on every category or just put all category images into one folder and run a inference on it what will be the accuracy ?

**Reply by rom1504** [reply]
Clip model is a general similarity model that works for any category as you can see in the demo https://rom1504.github.io/clip-retrieval/?back=https%3A%2F%2Fsplunk.vra.ro&index=laion_400m_128G
It provides a good similarity estimation between an image and a text.

Are you trying to implement a text search on a dataset of image ? If so then yes using a single image index can work.

What kind of query and results did you have exactly ?

Creating several indices can make sense if you know in advance the kind of items you want to retrieve but if you are working with a very small dataset of size 200, it doesn't make a lot of sense.

**Reply by Talhaz** [reply]
> Clip model is a general similarity model that works for any category as you can see in the demo https://rom1504.github.io/clip-retrieval/?back=https%3A%2F%2Fsplunk.vra.ro&index=laion_400m_128G
> It provides a good similarity estimation between an image and a text.
> 
> Are you trying to implement a text search on a dataset of image ? If so then yes using a single image index can work.
> 
> What kind of query and results did you have exactly ?
> 
> Creating several indices can make sense if you know in advance the kind of items you want to retrieve but if you are working with a very small dataset of size 200, it doesn't make a lot of sense.

I am working on a project which task is to first detect objects in an image and then search for similar images for that object. yeah Detector already classify object return a class from where it belongs, that's why it would be easier and accurate if I create indices of every category 
Yeah, images would be approximately for one category could be 3K I know it's not good .
_So My question is is it a good approach if I create different indices for every image category ??_

**Reply by rom1504** [reply]
Clip is probably not the best model for image-image similarity. You could try other models for example any classification trained model (see https://github.com/rom1504/image_embeddings ) or something like https://github.com/facebookresearch/dino

About creating one index per category you can do that yes. Another way is to post filter the results after the knn search

**Reply by Talhaz** [reply]
Thanks, I tried this one "https://github.com/rom1504/image_embeddings " In this, I didn't understand how to get similarities from a new input image because it runs only on its generated image embeddings, can u improve this functionality?

I will look into this too "https://github.com/facebookresearch/dino", thanks for ur help

**Reply by Talhaz** [reply]
https://github.com/rom1504/image_embeddings/issues/25   this is my question in that repo

**Reply by rom1504** [reply]
There, you could use efficientnet to generate the embedding of one image (https://github.com/qubvel/efficientnet)

You didn't explain yet clearly the problem you're having.
